File: LeetCodeRecord/MethodFirst/offer56-SingleNumbers.java
package MethodFirst;

class SingleNumbers {
    public int[] singleNumbers(int[] nums) {
        int ret = 0;
        for (int num : nums) {
            ret ^= num;
        }

        int div = 1;
        while ((ret & div) == 0) {
            div <<= 1;
        }

        int num1 = 0;
        int num2 = 0;
        for (int num : nums) {
            if ((num & div) == 0) {
                num1 ^= num;
            } else {
                num2 ^= num;
            }
        }

        return new int[]{num1, num2};
    }    

}

File: LeetCodeRecord/ACFirst-Golang/offer56-SingleNumbers.go
package main

func singleNumbers(nums []int) []int {
	if len(nums) == 0 {
		return []int{}
	}

	xorAll := 0
	for _, num := range nums {
		xorAll ^= num
	}

	index := 1
	for index & xorAll == 0 {
		index = index << 1
	}
	
	num1 := 0
	num2 := 0
	for _, num := range nums {
		if num & index == 0 {
			num1 ^= num
		} else {
			num2 ^= num
		}
	}

	return []int{num1, num2}
}